### DISCUSSION TITLE
Error: Builder returned invalid routes: should NOT be longer than 4096 characters

### DISCUSSION THREAD

**Discussion by chelsea-tnc** [vercel/vercel#8676] [Help] [answered]
We're trying to deploy but our build keeps failing with `Error: Builder returned invalid routes: should NOT be longer than 4096 characters.` We haven't made changes to the codebase nor our Vercel configurations since this started happening. 

We've tried:
- Running the build locally via `npm dev run.' It works fine.
- Re-deploying, but it still fails.
- Reviewing the rest of the logs to see what could be related. We're not seeing related issues or anything out of the ordinary.
- Reviewing the [documentation](https://vercel.com/docs/project-configuration) for hints on what could be the issue
- Reviewing [this](https://github.com/vercel/vercel/discussions/5940) issue for possible solutions but it doesn't look related. I made a comment where someone said they ran into a similar issue, but there isn't an update on how they resolved it. 

Does anyone know what we could troubleshoot next? Our deployment is in an error state and we're blocked.

**Comment by chelsea-tnc** [comment, marked as answer]
Update: after a 4th try, the deployment succeeded. I'm inclined to leave this discussion open, though, because we saw this issue twice already. Maybe someone has guidance on troubleshooting further when this happens again.

**Reply by amyegan** [reply]
There are a few [project configuration](https://vercel.com/docs/project-configuration) limits that you may be running into depending on your use case:
- [headers](https://vercel.com/docs/project-configuration#project-configuration/headers) can have a maximum of string length of 4,096 for the `source`, `key`, and `value` values.
- [redirects](https://vercel.com/docs/project-configuration#project-configuration/redirects) and [rewrites](https://vercel.com/docs/project-configuration#project-configuration/rewrites) can have a maximum of string length of 4,096 for the `source` and `destination` values.

If you are using any of those project config properties and have variable header, path, or query values, that could be the reason for occasional success. If something, such as a complex new token, is being generated at build time then it could sometimes cause the values to be over the limit. That is the first place I would look when debugging this kind of error.

**Reply by chelsea-tnc** [reply]
Thanks, @amyegan! I'll save your guidance in case I run into this issue again.
